### PR TITLE
Update the S3 named state handling to match consul

### PR DESCRIPTION
**Pull request by jbardin**
The S3 remote-state backend was using an env directory, which allowed
configurations with different state keys to list the same environments.
This turns out to have limited utility, and it's better to have matching
behavior between backends, so this changes the implementation to mirror
that of the consul backend.

### CODE REVIEW

**Comment by mitchellh** [comment]
Just curious what the difference is? i.e. what issues did the prior approach cause?

**Comment by jbardin** [comment]
The initial version prepended the key with a env directory at the root of the bucket. That allowed the environment list to be shared between clients using the same bucket with different keys. In hindsight it doesn't add much value, since there is no other shared data associated with the environment when using different keys, so matching the behavior of the builtin backends before releasing seemed more important.

**Comment by jbardin** [comment]
This PR missed the release today, and will now be a BC with the current release.

Should this be updated the automatically convert any existing backend envs?

**Comment by jbardin** [comment]
Leaving this here for now, but it needs an upgrade path.

If upgrading transparently during normal operation has too many corner cases, an failure then update during `init` may work.

**Comment by tomstockton** [comment]
I'm keen to use the state environment with S3 but slightly cautious of going ahead until this is merged.  I'm assuming it will be included in 0.9.3?  Any idea when this will be released?  Thanks.

**Comment by jbardin** [comment]
This turned into more code than I wanted for what started as a minor fix, but if this sufficiently covers the upgrade, we can remove it all after the next major release. 

When a legacy named state is detected, the Backend will return an `envUpgrader` which will move the state to the new path during the `Put` operation. The `envUpgrader` also manages moving the lock key to the new path and translating the outstanding lock ID. 

@mitchellh: let me know what you think

ACC tests:
=== RUN   TestOldEnvUpgrade
--- PASS: TestOldEnvUpgrade (20.27s)
	backend_test.go:118: creating S3 bucket terraform-remote-s3-test-58ebbca1 in us-east-1
=== RUN   TestOldEnvUpgradeLocked
--- PASS: TestOldEnvUpgradeLocked (17.43s)
	backend_test.go:118: creating S3 bucket terraform-remote-s3-test-58ebbcb5 in us-east-1
=== RUN   TestBackend_impl
--- PASS: TestBackend_impl (0.00s)
=== RUN   TestBackendConfig
--- PASS: TestBackendConfig (1.43s)
=== RUN   TestBackend
--- PASS: TestBackend (6.97s)
	backend_test.go:118: creating S3 bucket terraform-remote-s3-test-58ebbcc8 in us-east-1
=== RUN   TestBackendLocked
--- PASS: TestBackendLocked (17.55s)
	backend_test.go:118: creating S3 bucket terraform-remote-s3-test-58ebbccf in us-east-1
	testing.go:222: TestBackend: testing state locking for *s3.Backend
=== RUN   TestRemoteClient_impl
--- PASS: TestRemoteClient_impl (0.00s)
=== RUN   TestRemoteClient
--- PASS: TestRemoteClient (3.17s)
	backend_test.go:118: creating S3 bucket terraform-remote-s3-test-58ebbce0 in us-east-1
=== RUN   TestRemoteClientLocks
--- PASS: TestRemoteClientLocks (11.85s)
	backend_test.go:118: creating S3 bucket terraform-remote-s3-test-58ebbce3 in us-east-1
PASS

**Comment by jbardin** [comment]
Yes, this tests a combination of new old old paths in an s3 backend. I didn't see a clean way to attempt to migrate everything and still communicate this to the user if there were issues of some sort. I didn't want an network or permissions failure for an unrelated state to show up in a normal run. 

Maybe since 0.10 would require `init` anyways, we could add a hook in there and move each state at that point?

**Comment by mitchellh** [comment]
There is an interesting idea in there... maybe there is a special error return value for Configure or something that basically signals a re-initialize required. I'm imagning something like this:

```go
// in package "backend"
type ReinitError struct {
  Reason string
}
```

Then the "reason" is shown to the user as part of the normal "reason" field. This would let us return this on Configure if we have some old states or something. I guess that'd require a network request in Configure though which is not ideal...

I would prefer some way to just migrate users, and I think `init` is a good way to do so. That way we can just force and remove this as quickly as possible.

File: backend/remote-state/s3/backend_state.go [hunk 1]
```diff
@@ -147,7 +157,17 @@ func (b *Backend) path(name string) string {
 		return b.keyName
 	}
 
-	return strings.Join([]string{keyEnvPrefix, name, b.keyName}, "/")
+	return b.keyName + keyEnvPrefix + name
```
Comment (radeksimko) [inline backend/remote-state/s3/backend_state.go]: Shouldn't we store this with `.tfstate` extension just for clarity?

Comment (radeksimko) [inline backend/remote-state/s3/backend_state.go]: (possibly unrelated to this PR)
Probably a nitpick, but worth mentioning - the `key` attribute in the S3 `backend` block now has a different meaning, so it should probably be called `prefix` rather than `key`.

TL;DR this doesn't make sense anymore:

```hcl
backend "s3" {
    bucket  = "tf-test-bucket"
    key     = "something/terraform.tfstate"
    region  = "us-west-2"
  }
```

**Comment by jbardin** [comment]
Thanks @radeksimko, good point, the key is kind of a mixed meaning here. That _is_ the exact key for the "default" environment, extra environments add the env suffix. 

I think I'm going to drop this PR altogether. This was a initially a quick fix to match the consul behavior, but it's grown way too complex for my liking after becoming a BC. Since there's already bugs in flight for the existing layout, I think it would be easier to fix the existing layout rather than push on adding more and more moving parts. 

@mitchellh: I kind of like the idea of returning a `ReinitError` from Config, which could be useful in the future. I don't think network called from Config are a problem, the AWSClient already does that currently. The issue is that config errors are immediately fmt.Errorf'ed into the UI strings, outside of the initialization paths, so it's going to take a bit of work to make that happen.  All keep that in my TODO list for when we need it.
